rfc-106-docker-for-local-development.md

```diff
@@ -0,0 +1,41 @@
+# RFC 106: Use Docker for local development
+
+## Summary
+
+Adopt a Docker-based approach for local development instead of the Vagrant VM.
+
+## Problem
+
+We currently rely on the Vagrant development VM for the day to day dev cycle.
+
+- It has [many documented issues](https://docs.publishing.service.gov.uk/manual.html#development-vm)
+- Changes to puppet often result in a broken VM, because the infrastructure team don't use it.
+- We are adopting more cloud-based services into our stack, like S3 and Amazon's Elasticsearch. This means that our development environment doesn't look like production anymore, but we still need code in govuk-puppet.
+- It's hard to keep updated. Running the puppet command will often error and take a long time.
+- GOV.UK is moving to a containerised infrastructure
+
+## Possible approaches to local development
+
+- An approach based on **Docker and docker-compose**. Ben Thorner has a working prototype called [govuk-docker](https://github.com/benthorner/govuk-docker) and we have the [end-to-end tests](https://github.com/alphagov/publishing-e2e-tests) running in the same. Other GDS programmes like GOV.UK Pay use this approach.
+- An approach based on **GNU Guix** package manager, which Chris Baines uses via govuk-guix.
+- Because they are read-only, frontend applications on GOV.UK can be **run against production** using the startup.sh --live flag.
+- Developers on the GOV.UK PaaS work by developing against a **remote set of services** - we’d run many versions of GOV.UK in the cloud, which developers can use in development.
+- Applications can be run locally by **installing dependencies manually**. This is what some of GOV.UK’s Linux users do.
+
+
+Approach | Pro | Cons
+-- | -- | --
+Vagrant VM | Can look like production | Resource intensive, brittle
+Docker | Aligns with future hosting. Highly reproducible. | Slow because on a Mac it needs a VM. Development won’t match production exactly until we migrate Production onto a container-based platform.
+Guix | Huge amount of work put in | Technology unknown in GOV.UK
+Against production | Low setup, fast | Only works for frontend applications. Still needs local install
+Remote dependencies | Low setup. Can match Production exactly in terms of technologies used | Expensive, no offline development
+Manual install | Fast in development | Lots of setup
+
+## Proposal
+
+Adopt a Docker based local development environment.
+
+## Impact
+
+- How does local puppet development work?
```

Comment:
I try not to use the dev VM for running apps any more because it's a pain, so the main thing I use the dev VM for is for testing puppet changes.  It can be awkward to test those elsewhere:

- You can't test puppet locally outside the VM
- If you mess up a puppet change the easiest fix is often to rebuild the VM, which makes it a bit of a pain to test in integration (need to log in to the AWS console and terminate instances).

How will developing puppet work in the new world of docker?

Reply:
This is a really good question. What are some examples of puppet things you've tested in the VM?

Reply:
Here are a few:

- Getting both ES5 and ES6 running in CI so we can test search-api against both at the same time: https://github.com/alphagov/govuk-puppet/pull/9216

- Removing cronjobs: https://github.com/alphagov/govuk-puppet/pull/9015

- Tearing down rummager (particularly hard to test in integration!): https://github.com/alphagov/govuk-puppet/pull/9000

I think these will all be minor issues when production and CI are containerised and managed by something more reproducible than puppet: adding another container is straightforward; tearing stuff down should be no more complicated than just removing configuration; etc.  But until we're at that point, we need to do tedious things like manually tearing down state and managing cronjobs.

However, having said that: most of my puppet changes aren't really suitable for testing in the dev VM, stuff like modifying hostnames or environment-specific hieradata.  And I've made a fair few puppet changes which were *only* needed to keep the dev VM working.  So maybe this isn't a show-stopper, but for the changes which are easy to test locally and hard to test remotely the dev VM is nice.

Reply:
Will we also have two sets of configuration to manage? (i.e. Puppet vs docker world)

Reply:
@barrucadu thanks! Apart from the development VM, we used to be able to use `vagrant` to create specific VMs of a machine:

```sh
~/govuk/govuk-puppet $ vagrant up backend-1.backend
Bringing machine 'backend-1.backend' up with 'virtualbox' provider...
==> backend-1.backend: Box 'govuk_dev_trusty64_20160323' could not be found. Attempting to find and install...
    backend-1.backend: Box Provider: virtualbox
    backend-1.backend: Box Version: >= 0
==> backend-1.backend: Box file was not detected as metadata. Adding it directly...
==> backend-1.backend: Adding box 'govuk_dev_trusty64_20160323' (v0) for provider: virtualbox
```

Would that allow you to test those changes?

Reply:
@theseanything Yes, presumably. I'll add something to the impact section.

Reply:
@tijmenb That would be great

Reply:
Not sure if people are aware or if it's still true but the way we were told to work on puppet in the past was to write tests.  Our puppet repo has rspec in it and some prior art for testing it.

You can also spin up VMs for different classes of machines and then run puppet on them, which is how the old sysadmins used to do it.

How that translates to docker I don't know.

Reply:
I've been wondering about this myself. One thing I found valuable in my old job was a learning piece we did on 'how to write Puppet(/Ansible/Chef/Salt...) and provision a machine from scratch' - I have a feeling I'm not alone in my limited knowledge of Puppet. Somewhat outside the scope of this PR, I know, but I wonder if it would help support/clarify our testing approach we want to have for it?